File: wintest/ui/web/frontend/src/docs/click_element.ts
import type { StepDoc } from './types';

export const clickElementDoc: StepDoc = {
  name: 'click_element',
  title: 'Click Element',
  summary: 'Locate a UI element by natural-language description and click it.',
  description:
    'Uses the AI vision model to find the target element on the current screen, then clicks it. Useful when a UI layout may shift between runs. Slower and less deterministic than coordinate-based clicking; prefer the regular click step when the layout is stable. Requires the [ai] extras (torch, transformers).',
  parameters: [
    {
      name: 'target',
      type: 'string',
      required: true,
      description:
        'Natural language description of the UI element to click. Example: "Save button", "File menu".',
    },
    {
      name: 'click_type',
      type: 'string',
      required: false,
      description:
        'One of "click" (left, default), "double_click", "right_click", or "middle_click".',
    },
  ],
  example:
    '- action: click_element\n  target: "Save button"\n  description: "Save the file"',
};

File: wintest/steps/click_element.py
"""Click-element step -- locate a UI element by natural-language description and click it.

Uses the vision model to find the target on the current screen, then clicks
it. For clicks at known coordinates, use the regular click step instead.
"""

from ._base import StepDefinition, FieldDef


def validate(step, step_num):
    if not step.target:
        return [f"Step {step_num}: 'click_element' requires a 'target' field"]
    if step.click_type not in ("click", "double_click", "right_click", "middle_click"):
        return [f"Step {step_num}: invalid click_type '{step.click_type}'"]
    return []


def execute(step, agent):
    return agent.find_and_click(step, click_type=step.click_type or "click")


definition = StepDefinition(
    name="click_element",
    description="Locate a UI element by description and click it (uses the vision model)",
    fields=[
        FieldDef("target", "string", required=True),
        FieldDef("click_type", "string"),
    ],
    validate=validate,
    execute=execute,
    requires_vision=True,
)
